File: kafka-streams/src/test/groovy/io/micronaut/configuration/kafka/streams/KafkaTestInitializer.java
package io.micronaut.configuration.kafka.streams;

import io.micronaut.configuration.kafka.streams.optimization.OptimizationStream;
import io.micronaut.configuration.kafka.streams.startkafkastreams.StartKafkaStreamsOff;
import io.micronaut.configuration.kafka.streams.wordcount.WordCountStream;
import io.micronaut.context.annotation.BootstrapContextCompatible;
import io.micronaut.context.annotation.Value;
import io.micronaut.context.env.BootstrapPropertySourceLocator;
import io.micronaut.context.env.Environment;
import io.micronaut.context.env.PropertySource;
import io.micronaut.context.exceptions.ConfigurationException;
import jakarta.annotation.PostConstruct;
import jakarta.inject.Singleton;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.AdminClientConfig;
import org.apache.kafka.clients.admin.NewTopic;

import java.util.Collections;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@BootstrapContextCompatible
@Singleton
public class KafkaTestInitializer implements BootstrapPropertySourceLocator {

    private final Map<String, Object> adminProps;

    public KafkaTestInitializer(@Value("${kafka.bootstrap.servers}") String bootstrapServers) {
        this.adminProps = Collections.singletonMap(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
    }

    @PostConstruct
    void initializeTopics() {

        createTopics(Stream.of(WordCountStream.INPUT,
            WordCountStream.OUTPUT,
            WordCountStream.NAMED_WORD_COUNT_INPUT,
            WordCountStream.NAMED_WORD_COUNT_OUTPUT,
            StartKafkaStreamsOff.STREAMS_OFF_INPUT,
            StartKafkaStreamsOff.STREAMS_OFF_OUTPUT,
            OptimizationStream.OPTIMIZATION_ON_INPUT,
            OptimizationStream.OPTIMIZATION_OFF_INPUT).map(topicName -> configureTopic(topicName, 1, 1)).collect(Collectors.toSet()));
    }

    @Override
    public Iterable<PropertySource> findPropertySources(Environment environment) throws ConfigurationException {
        return BootstrapPropertySourceLocator.EMPTY_LOCATOR.findPropertySources(environment);
    }

    private NewTopic configureTopic(String name, int numPartitions, int replicationFactor) {
        return new NewTopic(name, numPartitions, (short) replicationFactor);
    }

    private void createTopics(Set<NewTopic> topicsToCreate) {
        try (AdminClient admin = AdminClient.create(adminProps)) {
            Set<String> existingTopics = admin.listTopics().names().get();
            Set<NewTopic> newTopics = topicsToCreate.stream().filter(newTopic -> !existingTopics.contains(newTopic.name())).collect(Collectors.toSet());
            admin.createTopics(newTopics).all().get();
        } catch (ExecutionException | InterruptedException e) {
            throw new IllegalStateException("Failed to initialize test kafka topics", e);
        }
    }
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/streams/KafkaTestInitializer.py
from typing import Annotated

from jakarta.inject import Singleton
from java.util import Collections
from micronaut.configuration.kafka.streams.event import BeforeKafkaStreamStart
from micronaut.context.annotation import Requires, Value
from micronaut.runtime.event.annotation import EventListener
from org.apache.kafka.clients.admin import AdminClient, AdminClientConfig, NewTopic


# The Java, Kotlin and Groovy suites create the topics from a BootstrapPropertySourceLocator; the
# bootstrap context is created before the GraalPy runtime, so the Python suite creates them right
# before the Kafka Streams are started instead.
@Requires(property="spec.name", value="WordCountStreamTest")
@Singleton
class KafkaTestInitializer:

    def __init__(self, bootstrap_servers: Annotated[str, Value("${kafka.bootstrap.servers}")]):
        self.admin_props = Collections.singletonMap(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrap_servers)

    @EventListener
    def initialize_topics(self, event: BeforeKafkaStreamStart) -> None:
        self.create_topics([
            self.configure_topic(topic_name, 1, 1)
            for topic_name in ["streams-plaintext-input", "named-word-count-input", "my-other-stream", "no-op-input"]
        ])

    @staticmethod
    def configure_topic(name: str, num_partitions: int, replication_factor: int) -> NewTopic:
        return NewTopic(name, num_partitions, replication_factor)

    def create_topics(self, topics_to_create: list[NewTopic]) -> None:
        admin = AdminClient.create(self.admin_props)
        try:
            existing_topics = admin.listTopics().names().get()
            new_topics = [new_topic for new_topic in topics_to_create if not existing_topics.contains(new_topic.name())]
            admin.createTopics(new_topics).all().get()
        finally:
            admin.close()
